staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -644,7 +648,9 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 	trace := utiltrace.New("cacher list", utiltrace.Field{Key: "type", Value: c.objectType.String()})
 	defer trace.LogIfLong(500 * time.Millisecond)
 
-	c.ready.wait()
+	if err := c.ready.wait(); err != nil {
+		return errors.NewServiceUnavailable(err.Error())
```

Comment:
I always lose track of what error has to be returned from the storage layer to trigger a 503 API error... did you verify this gets interpreted/propagated correctly by InterpretGetError/InterpretListError/InterpretWatchError

Reply:
if this bubbles correctly to the API, then lgtm

Reply:
I was thinking about it after submitting the PR meanwhile and I was going to test it locally, hold on

Reply:
ok, before doing more changes by intuition like I'm doing let me get this clear
Tt seems that we want `storage.IsUnreachable(err)` for Get and GetList so it retries (it is a 500 :shrug: )

https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/storage/errors/storage.go#L31-L32

https://github.com/kubernetes/kubernetes/blob/fbdd0d7b4165bc5a677d45e4dc693e3260297bfa/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L361-L376

and for the watch?

https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apiserver/pkg/storage/errors/storage.go#L106-L116

Reply:
hrmm... I'm fuzzy on the situations that lead us to return this error, and whether timeout or unavailable is what we want to communicate. @wojtek-t @sttts, do you have input on what is downstream of this that will react to the returned error, and which is more correct?

Reply:
If possible an integration test, e.g. k8s.io/apiserver/pkg/server/genericapiserver_test.go TestGracefulTermination would be good, to see that 50x really reaches the client.

@liggitt I guess in our case this is about clients not going through the load-balancer, but localhost (might be related to missing hairpinning support on some cloud platforms), that get stuck on shutdown of the local apiserver although they should connect to another master host.